### ISSUE TITLE
Doesn't build on macOS

### ISSUE THREAD

**Issue by wadetregaskis** [InternationalColorConsortium/DemoIccMAX#62]
iccProfLib uses C++11 features (e.g. `{}` syntax for object initialisation), but the projects use the Xcode default for C++ dialect, which is C++98.

iccProfLib also contains invalid code, that casts enums to integers and tries to use values outside their valid range (e.g. in `GetMeasurementFlareName`) which is undefined behaviour.  Clang errors on this by default, which can technically be silenced with `-Wno-enum-constexpr-conversion` although that's not a real solution (Clang doesn't guarantee the runtime behaviour will be as the code expects) and a performance pessimisation.

Beyond that, the iccProfLib builds as x86-64 _only_ but all the subprojects (e.g. IccApplyNamedCmm) use the default value for the target architectures (which is currently x86_64 and arm64).  So they all fail to link due to the missing arm64 slice.

RefIccMAXCmm also fails to build because it's using ColorSync types (e.g. `CMError`) which seemingly don't exist anymore (they were deprecated more than a decade ago, so I assume they've finally be outright removed from the framework).

The macOS deployment target is also being set to (or defaulting to?) 10.11, but modern Xcode doesn't support earlier than 10.13.  This nominally just results in a warning, although it's possible it's screwing other things up too (hard to tell with so many other errors occurring in parallel).

IccLibXML.xcodeproj is missing targets for the actual command line tools - IccToXML et al.

**Comment by xsscx** [comment]
Hello & Good Day-

A few comments and then some patches to get you compiled on macOS;

1. The Xcode Build is currently borked for all current macOS and Xcode version, I have a Patch coming via Pull Request soon.
2. Command Line cmake & make work fine, when a few Patches are applied.
3. IPatches for macOS based on Commit https://github.com/InternationalColorConsortium/DemoIccMAX/commit/4465cd3f2f2d4136c51a5f36ef1fba2e568aed9a

There are a few changes that need to be made to compile for macOS, here is a development config:

### Compiler Info
```
clang -v
Homebrew clang version 18.1.4
Target: x86_64-apple-darwin23.4.0
Thread model: posix
InstalledDir: /usr/local/opt/llvm/bin
```

### System Info
```
Sat May  4 09:56:14 EDT 2024
kern.version: Darwin Kernel Version 23.4.0: Fri Mar 15 00:11:05 PDT 2024; root:xnu-10063.101.17~1/RELEASE_X86_64
kern.osversion: 23E224
kern.iossupportversion: 17.4
kern.osproductversioncompat: 10.16
kern.osproductversion: 14.4.1
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk
Homebrew clang version 18.1.4
Target: x86_64-apple-darwin23.4.0
Thread model: posix
InstalledDir: /usr/local/opt/llvm/bin
```

## Patches for macOS Compile via Terminal based on Commit https://github.com/InternationalColorConsortium/DemoIccMAX/commit/4465cd3f2f2d4136c51a5f36ef1fba2e568aed9a
### Provided by @h02332 | DHOYT | SAT 04 MAY 2024 | macOS Compile Changes

##### Patch for DemoIccMAX-master/IccProfLib/icProfileHeader.h
```
diff -u ~/dmax/DemoIccMAX-master/IccProfLib/icProfileHeader.h ~/Downloads/DemoIccMAX-master/IccProfLib/icProfileHeader.h
--- /Users/xss/dmax/DemoIccMAX-master/IccProfLib/icProfileHeader.h	2024-05-04 09:47:49
+++ /Users/xss/Downloads/DemoIccMAX-master/IccProfLib/icProfileHeader.h	2023-11-13 12:55:59
@@ -1112,14 +1112,15 @@
  * Other enums
  */

+/** Measurement Flare, used in the measurmentType tag */
 typedef enum {
-    icFlare0 = 0x00000000,  // 0% flare
-    icFlare100 = 0x00000001,  // 100% flare
-    icMaxEnumFlareValid = 0x00000002  // Max valid flare, used for bounds in arrays or loops
+    icFlare0                            = 0x00000000,  /* 0% flare */
+    icFlare100                          = 0x00000001,  /* 100% flare */
 } icMeasurementFlare;

-#define icMaxEnumFlare ((icMeasurementFlare) 0xFFFFFFFF)  // For error handling or sentinel uses
-#define icMaxFlare     ((icMeasurementFlare) 0xFFFFFFFF)  // Same as above, ensure clarity in use
+/** Convenience Enum Definition - Not defined in ICC specification*/
+#define icMaxEnumFlare ((icMeasurementFlare) 0xFFFFFFFF)
+#define icMaxFlare     ((icMeasurementFlare) 0xFFFFFFFF) /* as defined by earlier versions */
```

##### Patch for DemoIccMAX-master/IccProfLib/IccUtil.h
```
diff -u ~/dmax/DemoIccMAX-master/IccProfLib/iccUtil.h ~/Downloads/DemoIccMAX-master/IccProfLib/IccUtil.h
--- /Users/xss/dmax/DemoIccMAX-master/IccProfLib/iccUtil.h	2024-05-04 09:35:21
+++ /Users/xss/Downloads/DemoIccMAX-master/IccProfLib/IccUtil.h	2023-11-13 12:55:59
@@ -73,7 +73,6 @@

 #include "IccDefs.h"
 #include "IccProfLibConf.h"
-#include <iostream>
 #include <string>
 #include <limits>

@@ -375,17 +374,17 @@
 class ICCPROFLIB_API CIccPixelBuf
 {
 public:
-  CIccPixelBuf(int nChan = icDefaultPixelBufSize);
+  CIccPixelBuf(int nChan=icDefaultPixelBufSize);
   ~CIccPixelBuf();
   icFloatNumber &operator[](int nPos) { return m_pixel[nPos]; }
-  icFloatNumber *get() { return m_pixel; }
+  icFloatNumber *get() { return m_pixel;}

   operator icFloatNumber *() { return m_pixel; }
   operator void *() { return m_pixel; }

 protected:
-  icFloatNumber m_buf[icDefaultPixelBufSize];  // Static buffer
-  icFloatNumber *m_pixel;  // Pointer to the buffer, either m_buf or dynamically allocated
+  icFloatNumber m_buf[icDefaultPixelBufSize];
+  icFloatNumber *m_pixel;
 };
```

##### Patch for DemoIccMAX-master/IccProfLib/iccUtil.cpp
```
diff -u ~/dmax/DemoIccMAX-master/IccProfLib/iccUtil.cpp ~/Downloads/DemoIccMAX-master/IccProfLib/IccUtil.cpp
--- /Users/xss/dmax/DemoIccMAX-master/IccProfLib/iccUtil.cpp	2024-05-04 09:50:22
+++ /Users/xss/Downloads/DemoIccMAX-master/IccProfLib/IccUtil.cpp	2023-11-13 12:55:59
@@ -1403,7 +1403,8 @@
           rng1.steps == rng2.steps);
 }

-CIccInfo::CIccInfo() : m_szStr(""), m_szSigStr("") {
+CIccInfo::CIccInfo() : m_szStr{}, m_szSigStr{}
+{
   m_str = new std::string;
 }

@@ -2053,19 +2054,17 @@

 const icChar *CIccInfo::GetMeasurementFlareName(icMeasurementFlare val)
 {
-  switch (val) { // Directly switch on the enum without casting to int
+  switch ((int)val) {
   case icFlare0:
-    return "Flare 0"; // Properly handle the 0% flare
+    return "Flare 0";

   case icFlare100:
-    return "Flare 100"; // Properly handle the 100% flare
+    return "Flare 100";

+  case icMaxEnumFlare:
+    return "Max Flare";
+
   default:
-    // Handle icMaxEnumFlare explicitly here if needed
-    if (val == icMaxEnumFlare) {
-      return "Max Flare"; // Special handling for the max sentinel value
-    }
-    // General default case for truly unexpected values
     sprintf(m_szStr, "Unknown Flare '%d'", (int)val);
     return m_szStr;
   }
@@ -2073,13 +2072,19 @@

 const icChar *CIccInfo::GetMeasurementGeometryName(icMeasurementGeometry val)
 {
-  switch (val) {  // Directly use enum type, casting is not needed
+  switch ((int)val) {
   case icGeometryUnknown:
     return "Geometry Unknown";
+
   case icGeometry045or450:
     return "Geometry 0-45 or 45-0";
+
   case icGeometry0dord0:
     return "Geometry 0-d or d-0";
+
+  case icMaxEnumGeometry:
+    return "Max Geometry";
+
   default:
     sprintf(m_szStr, "Unknown Geometry '%d'", (int)val);
     return m_szStr;
@@ -2591,20 +2596,20 @@
   return rv;
 }

-// Constructor with conditional memory allocation
-CIccPixelBuf::CIccPixelBuf(int nChan) : m_buf{} { // Initialize m_buf to zero if needed
-  if (nChan > icDefaultPixelBufSize) {
-    m_pixel = new icFloatNumber[nChan]; // Dynamically allocate memory if more is needed
-  } else {
-    m_pixel = m_buf; // Use the static buffer if sufficient
+CIccPixelBuf::CIccPixelBuf(int nChan/* =icDefaultPixelBufSize */) : m_buf{}
+{
+  if (nChan>icDefaultPixelBufSize) {
+    m_pixel = new icFloatNumber[nChan];
   }
+  else {
+    m_pixel = m_buf;
+  }
 }

-// Destructor with safe memory deallocation
-CIccPixelBuf::~CIccPixelBuf() {
-  if (m_pixel != m_buf) { // Only delete if m_pixel points to dynamically allocated memory
-    delete[] m_pixel;
-  }
+CIccPixelBuf::~CIccPixelBuf()
+{
+  if (m_pixel && m_pixel!=m_buf)
+    delete [] m_pixel;
 }

 #ifdef USEREFICCMAXNAMESPACE
```

### Compile Check
```
% cmake -DCMAKE_INSTALL_PREFIX=$HOME/.local -DCMAKE_BUILD_TYPE=Debug -DCMAKE_CXX_FLAGS="-g -fsanitize=address -fno-omit-frame-pointer -Wall -std=c++17" ./Build/Cmakecmake -DCMAKE_INSTALL_PREFIX=$HOME/.local -DCMAKE_BUILD_TYPE=Debug -DCMAKE_CXX_FLAGS="-g -fsanitize=address -fno-omit-frame-pointer -Wall -std=c++17" ./Build/Cmake
...
-- Generating done (0.1s)
-- Build files have been written to: ~/dmax/DemoIccMAX-master19
% make
[ 33%] Built target IccProfLib2
[ 66%] Built target IccProfLib2-static
[ 73%] Built target IccXML2
[ 81%] Built target IccXML2-static
[ 83%] Built target iccApplyNamedCmm
[ 84%] Built target iccDumpProfile
[ 86%] Built target iccRoundTrip
[ 89%] Built target iccApplyProfiles
[ 92%] Built target iccSpecSepToTiff
[ 95%] Built target iccTiffDump
[ 97%] Built target iccFromXml
[100%] Built target iccToXml
```


Let me know if any further compile issues.

**Comment by xsscx** [comment]
Patch PR: https://github.com/InternationalColorConsortium/DemoIccMAX/pull/64

**Comment by xsscx** [comment]
@wadetregaskis 

Wanted to confirm you can compile on macOS. 

PR https://github.com/InternationalColorConsortium/DemoIccMAX/pull/66 addressed the  error: integer value 4294967295 is outside the valid range of values [0, 3] for this enumeration type.

Update the Issue if can help.

**Comment by xsscx** [comment]
Closing this Issue as a workaround is provided and no activity for 30 days. If can help with any issues, please reopen or add new issue.
